src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs

```diff
@@ -76,20 +76,27 @@ public static OrderedSet<Symbol> Analyze(CSharpCompilation compilation, MethodSy
                 foreach (var kvp in lazyDisallowedCaptures)
                 {
                     var variable = kvp.Key;
-                    var type = (variable.Kind == SymbolKind.Local) ? ((LocalSymbol)variable).Type : ((ParameterSymbol)variable).Type;
 
-                    if (variable is SynthesizedLocal local && local.SynthesizedKind == SynthesizedLocalKind.Spill)
+                    if (variable is LocalSymbol local)
                     {
-                        Debug.Assert(local.TypeWithAnnotations.IsRestrictedType());
-                        diagnostics.Add(ErrorCode.ERR_ByRefTypeAndAwait, local.GetFirstLocation(), local.TypeWithAnnotations);
+                        if (local.SynthesizedKind is SynthesizedLocalKind.UserDefined or SynthesizedLocalKind.Spill)
+                        {
+                            reportLocalAcrossAwaitError(diagnostics, local, local.GetFirstLocation());
```

Comment:
>local.GetFirstLocation( [](http://example.com/codeflow?start=76&length=23)

Would it make sense to use first location from ```kvp.Value``` instead? Will it point to the actual problematic usage?